### PR TITLE
LOGIST-659 - Extract formatter

### PR DESCRIPTION
**Pull request by nenadstojanovikj**
This PR adds `GoogleCloudFormatter` which is used to format a Trace to be compatible with the Google Cloud `PATCH /traces` endpoint

### CODE REVIEW

File: src/GoogleCloudTracer/Recorder.php [hunk 1]
```diff
@@ -34,13 +35,21 @@ class Recorder implements RecorderInterface
...
      * @param AuthProviderInterface $authHandler
+     * @param TraceFormatterInterface $formatter
      * @param string $projectId
      */
-    public function __construct(AuthProviderInterface $authHandler, string $projectId)
+    public function __construct(AuthProviderInterface $authHandler, TraceFormatterInterface $formatter, string $projectId)
     {
         $this->authHandler = $authHandler;
         $this->projectId = $projectId;
+        $this->formatter = $formatter;
```
Comment (boekkooi-fresh) [inline src/GoogleCloudTracer/Recorder.php]: Could be allow formatter to be default?

File: tests/GoogleCloudTracer/Formatter/GoogleCloudFormatterTest.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    private $recorder;
+
+    public function setUp()
+    {
+        $this->traceId = 'trace-123';
+        $this->projectId = 'project-123';
+        $this->formatter = new GoogleCloudFormatter();
+
+        $this->recorder = new NoopRecorder();
```
Comment (boekkooi-fresh) [inline tests/GoogleCloudTracer/Formatter/GoogleCloudFormatterTest.php]: Could we not have this on the setup please.

File: tests/GoogleCloudTracer/Formatter/GoogleCloudFormatterTest.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * @var NoopRecorder
+     */
+    private $recorder;
+
+    public function setUp()
+    {
+        $this->traceId = 'trace-123';
+        $this->projectId = 'project-123';
```
Comment (boekkooi-fresh) [inline tests/GoogleCloudTracer/Formatter/GoogleCloudFormatterTest.php]: ...... really?

File: src/GoogleCloudTracer/Formatter/TraceFormatterInterface.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    /**
+     * Formats a trace to an arbitrary representation
+     *
+     * @param string $projectId
+     * @param string $traceId
+     * @param SpanInterface[] $spans
+     *
+     * @return mixed
+     */
+    public function formatTrace(string $projectId, string $traceId, array $spans);
```
Comment (boekkooi-fresh) [inline src/GoogleCloudTracer/Formatter/TraceFormatterInterface.php]: Should this not be split into formatSpan and formatTrace?

File: tests/GoogleCloudTracer/Formatter/GoogleCloudFormatterTest.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    public function checkTraceIsProperlyFormatted()
+    {
+        $recorder = new NoopRecorder();
+        $traceId = '11aa43be4634f4da';
+        $projectId = 'project-id';
+        $spanContext = new SpanContext($traceId, 1, false);
+        $span = new Span($recorder, microtime(true), 'trace-properly-formatted', $spanContext);
+        $span->finish();
+
+        $expectedData = json_decode(
```
Comment (boekkooi-fresh) [inline tests/GoogleCloudTracer/Formatter/GoogleCloudFormatterTest.php]: Why decode this?

**Comment by boekkooi-fresh** [comment]
:+1:

**Comment by Shru-10** [comment]
+1

**Comment by ogilliland** [comment]
+1

**Comment by emilyclaudet** [comment]
+1

**Pull request merged**